osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs

```diff
@@ -0,0 +1,304 @@
+using osu.Framework.Utils;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Configuration;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Objects.Drawables.Pieces;
+using osu.Game.Storyboards;
+using osuTK;
+using static osu.Game.Tests.Visual.OsuTestScene.ClockBackedTestWorkingBeatmap;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    [TestFixture]
+    public class TestSceneSliderSnaking : TestSceneOsuPlayer
+    {
+        [Resolved]
+        private AudioManager audioManager { get; set; }
+
+        private TrackVirtualManual track;
+
+        protected override bool Autoplay => autoplay;
+        private bool autoplay;
+
+        private readonly Bindable<bool> snakingIn = new Bindable<bool>();
+        private readonly Bindable<bool> snakingOut = new Bindable<bool>();
+
+        protected override WorkingBeatmap CreateWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null)
+        {
+            var working = new ClockBackedTestWorkingBeatmap(beatmap, storyboard, new FramedClock(new ManualClock { Rate = 1 }), audioManager);
+            track = (TrackVirtualManual)working.Track;
+            return working;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(RulesetConfigCache configCache)
+        {
+            var config = (OsuRulesetConfigManager)configCache.GetConfigFor(Ruleset.Value.CreateInstance());
+            config.BindWith(OsuRulesetSetting.SnakingInSliders, snakingIn);
+            config.BindWith(OsuRulesetSetting.SnakingOutSliders, snakingOut);
+        }
+
+        private DrawableSlider slider;
+        private DrawableSliderRepeat repeat;
+        private Vector2 vector;
+
+        [SetUpSteps]
+        public override void SetUpSteps() { }
+
+        [Test]
+        public void TestSnaking()
+        {
+            AddStep("have autoplay", () => autoplay = true);
+            base.SetUpSteps();
+            AddUntilStep("wait for track to start running", () => track.IsRunning);
+
+            AddStep("retrieve 1st slider", () => slider = (DrawableSlider)Player.DrawableRuleset.Playfield.AllHitObjects.First());
+            testLinear(true);
+            testLinear(false);
+            AddStep("retrieve 2nd slider", () => slider = (DrawableSlider)Player.DrawableRuleset.Playfield.AllHitObjects.Skip(1).First());
+            testRepeating(true);
+            testRepeating(false);
+            AddStep("retrieve 3rd slider", () => slider = (DrawableSlider)Player.DrawableRuleset.Playfield.AllHitObjects.Skip(2).First());
+            testDoubleRepeating(true);
+            testDoubleRepeating(false);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public void TestArrowStays(bool isHit)
+        {
+            var isSame = isHit ? "is same" : "decreased";
+            var enable = isHit ? "enable" : "disable";
```

Comment:
These are named super confusingly (having `isSame` not have a value of `"is same"` will trip people up).

Reply:
I didn't know how to include strings in formatted strings (or what ever they are called in c#) so I took them outside. These are only needed for step descriptions and you could just put them in the string to begin with.

Reply:
You can nest ternaries in interpolated strings but I'm not necessarily advocating for that, just better naming that will encompass both values of the strings (like `repeatVectorState` and `snakingSettingState` or something).

Reply:
I went with nested ternaries anyway. Shouldn't be that big of a deal now that the code is pretty DRY.